ome/omero-iviewer
# Quantization family only if needed

This attempts to deal with the issue of long query strings for ```render_image``` if we add quantization settings to every channel, as described at https://github.com/ome/omero-iviewer/pull/144#issuecomment-378880392

Initial (simple) commit simply omitted the ```quantization``` from the ```maps``` if the family was "linear".
E.g. for a single channel image with "linear" family:
```
?maps=[{"inverted":{"enabled":false}]
```
However, in the case that this image has *saved* settings with a non-linear quantization, this would render the saved settings (instead of "linear"). 

So, we need to know when the current settings are different from the saved settings, so we can add the 'linear' family when the saved quantization is not linear.
```
maps=[{"inverted":{"enabled":false},"quantization":{"family":"linear"}}
```
We can simply make a copy of the ```channels_info_``` when the ```ol3 Image``` is created, BUT we also need to know when the user hits Save.
This required a new Event, triggered when the user hits Save, to update the ```saved_channel_info``` in the ol3 Image.

NB: we omit the "coefficient" when family is "linear" to avoid bug in webgateway that may not be fixed by the time that iviewer is released https://github.com/openmicroscopy/openmicroscopy/pull/5740 . However, this doesn't fix the coefficient range limited to 0-1 which is also fixed by that PR. NB: I'll close that PR while this one is tested, so we can check the behaviour of iviewer with current OMERO.

To test:
 - Open iviewer showing an image that has small number of channels (easier to see what's going on).
 - Open browser devtools, click Network tab and choose Img to see what image URLs are being requested.
 - When rendering setting changes are made, check the URL of rendered images. Note that the quantization is missing from the ```?map``` (see above).
 - Choose to set quantization family of Polynomial and adjust gamma. Image should be rendered correctly and URL should include "polynomial" setting and "coefficient".
 - Revert back to "linear" and the URL should revert to not include quantization. Also check the rendered image reverts to linear appearance.
 - Now switch to "polynomial" again and Save settings.
 - Then revert to "linear again". This time, the URL *will* include quantization "linear", but this shouldn't include "coefficient". Rendered image should appear correct.
 - Check that this works with multiple viewers (with *same* Image) updating correctly on Save
    - Open an Image with "linear" quantization in 2 viewers
    - In one viewer, choose "polynomial" and Save. Both viewers update.
    - In the other viewer, revert to "linear" and check that this is displayed correctly.

NB: I noticed that ```Save To All``` is ignoring quantization, so this will need fixing in another PR (webgateway). Can workaround by Save first, then Save to All.
